csrc/preseg_passes/make_resharding_contiguous.cpp

```diff
@@ -109,7 +111,27 @@ void setLoopAndAllocationDomain(TensorView* tv) {
       " as ",
       alloc_dom);
   tv->reorder(permutation.value());
-  tv->setAllocationDomain(tv->getLoopDomain(), contiguity);
+
+  if (is_resharding) {
+    // Resharding expressions have specific requirements on position of
+    // gathered/scattered dimensions in the allocation domain that is ensured
+    // by ReorderShardedAxisPass. So we do not move the DIDx to the front in
+    // this case. For example, in reduce-scatter, the scattered axis is the
+    // outer-most dimension in communication input and output.
+    tv->setAllocationDomain(tv->getLoopDomain(), contiguity);
+    return;
+  }
+
+  // Most schedulers require DIDx to be at the front of the loop domain.
+  auto old2new = reorderDIDToFront(tv);
+  auto new2old = ir_utils::normalizeOld2New(old2new, tv->nDims());
+  std::vector<std::optional<bool>> reordered_contiguity;
+  std::transform(
+      new2old.begin(),
+      new2old.end(),
+      std::back_inserter(reordered_contiguity),
+      [contiguity](int64_t i) -> std::optional<bool> { return contiguity[i]; });
+  tv->setAllocationDomain(tv->getLoopDomain(), reordered_contiguity);
```

Comment:
This section appears to be bandaids. What should happen here when #4381 is properly done?

Reply:
Loop domain:
- Reorder as logical domain (instead of allocation domain currently) with DIDx at the front.

Allocation domain:
- We should not need to move DIDx to the front. The allocation domain should inherit the transforms and shardings as loop domain, with the exception of ParallelType::Stream where loop and allocation differ. We likely need a more comprehensive propagation for ParallelType::Stream since it may or may not be present in allocation domain.
- Since we do not need to reorder the DIDx, the contiguity does not need any reordering either (L127-L133).

Reply:
SG. I'll skip reviewing the bandaids here, looking forward to the next PR to clean it up.